File: wrappers/js/test/devkit-helper.js
/**
 * Yaci DevKit HTTP helper for integration tests.
 * Uses Bun's built-in fetch API.
 *
 * Every request carries an AbortSignal timeout so a wedged devnet fails the call fast (with a
 * clear TimeoutError) instead of hanging until the test's whole timeout budget is gone.
 */

const DEVKIT_URL = "http://localhost:10000/local-cluster/api";

// General bound for chain-data calls. The reset POST gets a larger bound because DevKit 0.12's
// reset handler blocks while it re-bootstraps the cluster (~20-30s when healthy).
const REQUEST_TIMEOUT_MS = 30_000;
const RESET_TIMEOUT_MS = 60_000;
// How long reset() polls for the devnet to serve chain data again before re-POSTing the reset.
const HEALTH_BUDGET_MS = 60_000;
const RESET_ATTEMPTS = 3;

export class DevKitHelper {
  constructor(baseUrl = DEVKIT_URL) {
    this.baseUrl = baseUrl;
  }

  // reset restarts the devnet and returns only once it serves chain data again.
  //
  // DevKit 0.12 (companion mode) re-bootstraps the whole cluster on reset, and that bootstrap can
  // wedge (e.g. the relay never syncs from the companion within its window), leaving the node
  // socket dead until the NEXT reset POST kicks the cluster back to life. So: POST the reset, poll
  // until the chain-data API answers, and if the devnet stays dead re-POST the reset.
  async reset() {
    let lastErr;
    for (let attempt = 1; attempt <= RESET_ATTEMPTS; attempt++) {
      try {
        await fetch(`${this.baseUrl}/admin/devnet/reset`, {
          method: "POST",
          signal: AbortSignal.timeout(RESET_TIMEOUT_MS),
        });
      } catch (e) {
        // The bootstrap keeps running server-side; the health poll below is what decides.
        lastErr = e;
      }
      if (await this.waitHealthy(HEALTH_BUDGET_MS)) return;
      lastErr = new Error("devnet did not serve chain data after reset");
      console.log(`devkit reset attempt ${attempt}/${RESET_ATTEMPTS}: devnet still down, re-posting reset`);
    }
    throw new Error(`devnet reset failed after ${RESET_ATTEMPTS} attempts: ${lastErr}`);
  }

  // Healthy = the chain-data API (yaci-store, fed by the node) answers with protocol parameters.
  // /admin/devnet alone is no proof: it stays 200 while the node socket is dead.
  async waitHealthy(budgetMs) {
    const deadline = Date.now() + budgetMs;
    while (Date.now() < deadline) {
      await new Promise((r) => setTimeout(r, 3000));
      try {
        const resp = await fetch(`${this.baseUrl}/epochs/parameters`, {
          signal: AbortSignal.timeout(5000),
        });
        if (resp.ok) {
          await resp.json();
          return true;
        }
      } catch {
        // keep polling
      }
    }
    return false;
  }

  async topup(address, adaAmount = 100) {
    // reset() already health-gates the devnet, but the faucet can still transiently refuse right
    // after the hand-over to the node. Retry with backoff.
    let lastErr;
    for (let attempt = 1; attempt <= 8; attempt++) {
      try {
        const resp = await fetch(`${this.baseUrl}/addresses/topup`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ address, adaAmount }),
          signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
        });
        const result = await resp.json();
        if (resp.ok && !(result && result.status === false)) return result;
        lastErr = new Error(`topup failed (${resp.status}): ${JSON.stringify(result)}`);
      } catch (e) {
        lastErr = e;
      }
      await new Promise((r) => setTimeout(r, 4000));
    }
    throw lastErr;
  }

  async getUtxos(address) {
    const resp = await fetch(`${this.baseUrl}/addresses/${address}/utxos`, {
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
    });
    return resp.json();
  }

  async getProtocolParams() {
    const resp = await fetch(`${this.baseUrl}/epochs/parameters`, {
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
    });
    return resp.json();
  }

  // After a reset, the devkit's backend submit-api (port 8090) can lag behind the chain-data API
  // that reset() health-gates on — the devkit then returns 400 wrapping "Connection refused".
  // That's the devnet still booting, not a ledger rejection, so retry it; genuine rejections
  // surface immediately.
  async submitTx(txCborHex) {
    const txBytes = Buffer.from(txCborHex, "hex");
    let lastText;
    for (let attempt = 1; attempt <= 8; attempt++) {
      const resp = await fetch(`${this.baseUrl}/tx/submit`, {
        method: "POST",
        headers: { "Content-Type": "application/cbor" },
        body: txBytes,
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });
      const text = await resp.text();
      if (resp.ok || !text.includes("Connection refused")) {
        return text.trim().replace(/"/g, "");
      }
      lastText = text;
      await new Promise((r) => setTimeout(r, 4000));
    }
    return lastText.trim().replace(/"/g, "");
  }

  async getTx(txHash) {
    const resp = await fetch(`${this.baseUrl}/txs/${txHash}`, {
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
    });
    return resp.json();
  }

  async waitForBlock(ms = 2000) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  async isAvailable() {
    try {
      const resp = await fetch(`${this.baseUrl}/admin/devnet`, {
        signal: AbortSignal.timeout(3000),
      });
      return resp.status === 200;
    } catch {
      return false;
    }
  }
}

File: wrappers/python/tests/devkit_helper.py
"""Yaci DevKit HTTP helper for integration tests.

Uses only stdlib (urllib) - no external dependencies required.
"""
import json
import time
import urllib.request
import urllib.error

DEVKIT_URL = "http://localhost:10000/local-cluster/api"


class DevKitHelper:
    """Helper class for interacting with Yaci DevKit local cluster."""

    def __init__(self, base_url=DEVKIT_URL):
        self.base_url = base_url

    def reset(self):
        """Reset the devnet and return only once it serves chain data again.

        DevKit 0.12 (companion mode) re-bootstraps the whole cluster on reset, and that bootstrap
        can wedge (e.g. the relay never syncs from the companion within its window), leaving the
        node socket dead until the NEXT reset POST kicks the cluster back to life. So: POST the
        reset, poll until the chain-data API answers, and if the devnet stays dead re-POST it.
        """
        last_err = None
        for attempt in range(1, 4):
            req = urllib.request.Request(
                f"{self.base_url}/admin/devnet/reset",
                method="POST",
                data=b"",
            )
            try:
                # The reset handler blocks while the cluster re-bootstraps (~20-30s when healthy).
                with urllib.request.urlopen(req, timeout=60) as resp:
                    status = resp.status
            except (urllib.error.URLError, OSError, TimeoutError) as e:
                # The bootstrap keeps running server-side; the health poll below decides.
                last_err = e
                status = None
            if self._wait_healthy(60):
                return status
            last_err = RuntimeError("devnet did not serve chain data after reset")
            print(f"devkit reset attempt {attempt}/3: devnet still down, re-posting reset")
        raise RuntimeError(f"devnet reset failed after 3 attempts: {last_err}")

    def _wait_healthy(self, budget_seconds):
        """Poll until the chain-data API (yaci-store, fed by the node) answers with parameters.

        /admin/devnet alone is no proof: it stays 200 while the node socket is dead.
        """
        deadline = time.monotonic() + budget_seconds
        while time.monotonic() < deadline:
            time.sleep(3)
            try:
                url = f"{self.base_url}/epochs/parameters"
                with urllib.request.urlopen(url, timeout=5) as resp:
                    if resp.status == 200:
                        json.loads(resp.read())
                        return True
            except (urllib.error.URLError, OSError, ValueError):
                continue
        return False

    def topup(self, address, ada_amount=100):
        """Fund an address with ADA.

        reset() already health-gates the devnet, but the faucet can still transiently fail right
        after the hand-over to the node. Retry with backoff.
        """
        data = json.dumps({"address": address, "adaAmount": ada_amount}).encode()
        last_err = None
        for _ in range(8):
            req = urllib.request.Request(
                f"{self.base_url}/addresses/topup",
                method="POST",
                data=data,
                headers={"Content-Type": "application/json"},
            )
            try:
                with urllib.request.urlopen(req, timeout=30) as resp:
                    result = json.loads(resp.read())
                if not (isinstance(result, dict) and result.get("status") is False):
                    return result
                last_err = RuntimeError(f"topup failed: {result}")
            except urllib.error.HTTPError as e:
                last_err = RuntimeError(
                    f"topup failed: HTTP {e.code}: {e.read().decode('utf-8', 'replace')}")
            time.sleep(4)
        raise last_err

    def get_utxos(self, address):
        """Fetch UTXOs for an address."""
        url = f"{self.base_url}/addresses/{address}/utxos"
        with urllib.request.urlopen(url, timeout=30) as resp:
            return json.loads(resp.read())

    def get_protocol_params(self):
        """Fetch current protocol parameters."""
        url = f"{self.base_url}/epochs/parameters"
        with urllib.request.urlopen(url, timeout=30) as resp:
            return json.loads(resp.read())

    def submit_tx(self, tx_cbor_hex):
        """Submit a signed transaction (CBOR hex string).

        Converts hex to raw bytes and POSTs as application/cbor.

        After a reset, the devkit's backend submit-api (port 8090) can lag behind the chain-data
        API that reset() health-gates on — the devkit then returns 400 wrapping "Connection
        refused". That's the devnet still booting, not a ledger rejection, so retry it; genuine
        rejections surface immediately.
        """
        tx_bytes = bytes.fromhex(tx_cbor_hex)
        last_err = None
        for _ in range(8):
            req = urllib.request.Request(
                f"{self.base_url}/tx/submit",
                method="POST",
                data=tx_bytes,
                headers={"Content-Type": "application/cbor"},
            )
            try:
                with urllib.request.urlopen(req, timeout=30) as resp:
                    return resp.read().decode("utf-8").strip().strip('"')
            except urllib.error.HTTPError as e:
                body = e.read().decode("utf-8", "replace")
                if "Connection refused" not in body:
                    raise RuntimeError(f"tx submit failed: HTTP {e.code}: {body}") from None
                last_err = RuntimeError(f"tx submit failed: HTTP {e.code}: {body}")
            time.sleep(4)
        raise last_err

    def get_tx(self, tx_hash):
        """Get transaction details by hash."""
        url = f"{self.base_url}/txs/{tx_hash}"
        with urllib.request.urlopen(url, timeout=30) as resp:
            return json.loads(resp.read())

    def get_latest_epoch(self):
        """Current epoch, for intents whose certificates carry epoch bounds (e.g. pool retirement).

        Prefers the protocol-params response (Blockfrost-style params carry "epoch"), falls back
        to the Blockfrost-compatible /epochs/latest.
        """
        pp = self.get_protocol_params()
        epoch = pp.get("epoch")
        if isinstance(epoch, int):
            return epoch
        if isinstance(epoch, str):
            return int(epoch)
        url = f"{self.base_url}/epochs/latest"
        with urllib.request.urlopen(url, timeout=30) as resp:
            latest = json.loads(resp.read())
        return int(latest["epoch"])

    def wait_for_block(self, seconds=2):
        """Wait for a new block to be produced."""
        time.sleep(seconds)

    def is_available(self):
        """Check if DevKit is running."""
        try:
            req = urllib.request.Request(
                f"{self.base_url}/admin/devnet",
                method="GET",
            )
            with urllib.request.urlopen(req, timeout=3) as resp:
                return resp.status == 200
        except (urllib.error.URLError, OSError):
            return False
